File: DropDownTableViewObjC/DropDownTableViewObjC/TableViewController.m
//
//  TableViewController.m
//  DropDownTableViewObjC
//
//  Created by NSSimpleApps on 15.01.17.
//  Copyright © 2017 NSSimpleApps. All rights reserved.
//

#import "TableViewController.h"
#import "UIAlertController+TextField.h"

@interface Pair : NSObject

@property (copy) NSString *key;
@property (copy) NSString *value;

- (instancetype)initWithKey:(NSString *)key value:(NSString *)value;

@end

@implementation Pair

- (instancetype)initWithKey:(NSString *)key value:(NSString *)value {
    
    self = [super init];
    
    if (self) {
        
        self.key = key;
        self.value = value;
    }
    
    return self;
}

@end

@interface State : NSObject

@property (copy) NSString *name;
@property (strong) NSMutableOrderedSet<Pair *> *parameters;

@end

@implementation State

- (instancetype)init {
    
    self = [super init];
    
    if (self) {
        
        self.name = @"";
        self.parameters = [NSMutableOrderedSet orderedSet];
    }
    
    return self;
}

@end

@interface TableViewController ()

@property (strong) NSMutableArray<State *> *data;

@end

@implementation TableViewController

- (void)viewDidLoad {
    
    [super viewDidLoad];
    
    State *russia = [State new];
    russia.name = @"Russia";
    [russia.parameters addObject:[[Pair alloc] initWithKey:@"Capital" value:@"Moscow"]];
    [russia.parameters addObject:[[Pair alloc] initWithKey:@"Square" value:@"17 098 242"]];
    
    State *usa = [State new];
    usa.name = @"USA";
    [usa.parameters addObject:[[Pair alloc] initWithKey:@"Capital" value:@"Washington, D.C."]];
    [usa.parameters addObject:[[Pair alloc] initWithKey:@"Square" value:@"9 857 306"]];
    [usa.parameters addObject:[[Pair alloc] initWithKey:@"Independence Day" value:@"July 4"]];
    
    State *zimbabwe = [State new];
    zimbabwe.name = @"Zimbabwe";
    [zimbabwe.parameters addObject:[[Pair alloc] initWithKey:@"Capital" value:@"Harare"]];
    
    self.data = @[russia, usa, zimbabwe].mutableCopy;
    
    self.navigationItem.leftBarButtonItem = self.editButtonItem;
}

//- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
//    
//    return 1;
//}
//
//- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
//    
//    return self.data.count;
//}
//
//- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
//    
//    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"RowCell"
//                                                            forIndexPath:indexPath];
//    cell.textLabel.text = self.data[indexPath.row].name;
//    
//    return cell;
//}

- (NSInteger)numberOfRowsInTableView:(UITableView *)tableView {
    
    return self.data.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfSubrowsInRow:(NSInteger)row {
    
    return self.data[row].parameters.count;
}

- (NSNumber *)showSubrowsInRow {
    
    return @0;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRow:(NSInteger)row indexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"RowCell"
                                                            forIndexPath:indexPath];
    cell.textLabel.text = self.data[row].name;
    
    return cell;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForSubrow:(NSInteger)subrow inRow:(NSInteger)row indexPath:(NSIndexPath *)indexPath {
    
    Pair *pair = [self.data[row].parameters objectAtIndex:subrow];
    
    UITableViewCell *cell =
    [tableView dequeueReusableCellWithIdentifier:@"SubrowCell"
                                    forIndexPath:indexPath];
    cell.textLabel.text = pair.key;
    cell.detailTextLabel.text = pair.value;
    
    return cell;
}

- (UIView *)tableView:(UITableView *)tableView accessoryViewForSelectedRow:(NSInteger)row {
    
    return [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"selectedImage"]];
}

- (UIView *)tableView:(UITableView *)tableView accessoryViewForDeselectedRow:(NSInteger)row {
    
    return [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"deselectedImage"]];
}

- (NSInteger)tableView:(UITableView *)tableView indentationLevelForRow:(NSInteger)row {
    
    return 0;
}

- (NSInteger)tableView:(UITableView *)tableView indentationLevelForSubrow:(NSInteger)subrow inRow:(NSInteger)row {
    
    return 3;
}

- (BOOL)tableView:(UITableView *)tableView canEditRow:(NSInteger)row {
    
    return YES;
}

- (BOOL)tableView:(UITableView *)tableView canEditSubrow:(NSInteger)subrow inRow:(NSInteger)row {
    
    return YES;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRow:(NSInteger)row {
    
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        
        [self.data removeObjectAtIndex:row];
        [tableView deleteRowsAtRows:@[@(row)] withRowAnimation:UITableViewRowAnimationAutomatic];
        
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        
        
    }
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forSubrow:(NSInteger)subrow inRow:(NSInteger)row {
    
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        
        [self.data[row].parameters removeObjectAtIndex:subrow];
        [tableView deleteSubrows:@[@(subrow)] inRow:row withRowAnimation:UITableViewRowAnimationAutomatic];
        
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        
        UIAlertController *ac =
        [UIAlertController alertControllerWithTitle:@"Insert new subrow"
                                       placeholders:@[@"parameter", @"value"]
                                         completion:^(NSArray<NSString *> * _Nonnull texts) {
                                             
                                             NSString *key = texts.firstObject;
                                             NSString *value = texts.lastObject;
                                             
                                             [self.data[row].parameters insertObject:[[Pair alloc] initWithKey:key value:value]
                                                                             atIndex:0];
                                             [tableView insertSubrows:@[@0] inRow:row withRowAnimation:UITableViewRowAnimationAutomatic];
                                         }];
        
        [self presentViewController:ac animated:YES completion:nil];
    }
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRow:(NSInteger)row {
    
    return UITableViewCellEditingStyleDelete;
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForSubrow:(NSInteger)subrow inRow:(NSInteger)row {
    
    return UITableViewCellEditingStyleDelete;
}

- (void)tableView:(UITableView *)tableView didEndDisplayingCell:(UITableViewCell *)cell forRow:(NSInteger)row {
    
    NSLog(@"%@, %ld", cell.textLabel.text, (long)row);
}

- (void)tableView:(UITableView *)tableView didEndDisplayingCell:(UITableViewCell *)cell forSubrow:(NSInteger)subrow inRow:(NSInteger)row {
    
    NSLog(@"%@, %ld, %ld", cell.detailTextLabel.text, (long)row, (long)subrow);
}

- (BOOL)tableView:(UITableView *)tableView canMoveRow:(NSInteger)row {
    
    return YES;
}

- (BOOL)tableView:(UITableView *)tableView canMoveSubrow:(NSInteger)subrow inRow:(NSInteger)row {
    
    return YES;
}

- (NSInteger)tableView:(UITableView *)tableView targetRowForMoveFromRow:(NSInteger)sourceRow toProposedDestinationRow:(NSInteger)destinationRow {
    
    return destinationRow;
}

- (NSInteger)tableView:(UITableView *)tableView targetSubrowForMoveFromSubrow:(NSInteger)sourceSubrow toProposedDestinationSubrow:(NSInteger)destinationSubrow inRow:(NSInteger)row {
    
    return destinationSubrow;
}

- (void)tableView:(UITableView *)tableView moveRow:(NSInteger)sourceRow toDestinationRow:(NSInteger)destinationRow {
    
    [self.data exchangeObjectAtIndex:sourceRow withObjectAtIndex:destinationRow];
}

- (void)tableView:(UITableView *)tableView moveSourceSubrow:(NSInteger)sourceSubrow toDestinationSubrow:(NSInteger)destinationSubrow inRow:(NSInteger)row {
    
    State *state = self.data[row];
    [state.parameters exchangeObjectAtIndex:sourceSubrow withObjectAtIndex:destinationSubrow];
}

- (UITableViewRowAnimation)tableView:(UITableView *)tableView animationForInsertionInRow:(NSInteger)row {
    
    return UITableViewRowAnimationRight;
}

- (UITableViewRowAnimation)tableView:(UITableView *)tableView animationForDeletionInRow:(NSInteger)row {
    
    return UITableViewRowAnimationLeft;
}

@end

File: DropDownTableViewSwift/DropDownTableViewSwift/TableViewController.swift
//
//  TableViewController.swift
//  DropDownTableView
//
//  Created by NSSimpleApps on 04.11.15.
//  Copyright © 2015 NSSimpleApps. All rights reserved.
//

import UIKit

class Pair: NSObject {
    
    var key = ""
    var value = ""
    
    init(key: String, value: String) {
        
        self.key = key
        self.value = value
    }
}

class State: NSObject {
    
    var name = ""
    
    var parameters = NSMutableOrderedSet()
}

class TableViewController: DropDownTableViewController {
    
    var data: [State] = []
    
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        let russia = State()
        russia.name = "Russia"
        russia.parameters.add(Pair(key: "Capital", value: "Moscow"))
        russia.parameters.add(Pair(key: "Square", value: "17 098 242"))
        
        let usa = State()
        usa.name = "USA"
        usa.parameters.add(Pair(key: "Capital", value: "Washington, D.C."))
        usa.parameters.add(Pair(key: "Square", value: "9 857 306"))
        usa.parameters.add(Pair(key: "Independence Day", value: "July 4"))
        
        let zimbabwe = State()
        zimbabwe.name = "Zimbabwe"
        zimbabwe.parameters.add(Pair(key: "Capital", value: "Harare"))
        
        self.data = [russia, usa, zimbabwe]
        
        self.navigationItem.leftBarButtonItem = self.editButtonItem
    }
    
    override func numberOfRows(in tableView: UITableView) -> Int {
        
        return self.data.count
    }
    
    override func tableView(_ tableView: UITableView, numberOfSubrowsInRow row: Int) -> Int {
        
        return self.data[row].parameters.count
    }
    
    override var showSubrowsInRow: Int? {
        
        return 1
    }
    
    override func tableView(_ tableView: UITableView, cellForRow row: Int, indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "RowCell", for: indexPath)
        cell.textLabel?.text = self.data[row].name
        
        if row == self.nsk_selectedRow {
            
            cell.accessoryView = UIImageView(image: UIImage(named: "up_arrow"))
            
        } else {
            
            cell.accessoryView = UIImageView(image: UIImage(named: "down_arrow"))
        }
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRow row: Int) {
        
        switch (self.nsk_selectedRow, row) {
            
        case (let sr?, _) where row == sr:
            tableView.cellForRow(at: row)?.accessoryView =  UIImageView(image: UIImage(named: "down_arrow"))
            tableView.deselect(row: row, animated: true)
            break
            
        case (let sr?, _) where row != sr:
            tableView.cellForRow(at: row)?.accessoryView = UIImageView(image: UIImage(named: "up_arrow"))
            tableView.cellForRow(at: sr)?.accessoryView = UIImageView(image: UIImage(named: "down_arrow"))
            break
            
        case (nil, _):
            tableView.cellForRow(at: row)?.accessoryView = UIImageView(image: UIImage(named: "up_arrow"))
            break
            
        default:
            break
        }
    
        super.tableView(tableView, didSelectRow: row)
    }
    
    override func tableView(_ tableView: UITableView, cellForSubrow subrow: Int, inRow row: Int, indexPath: IndexPath) -> UITableViewCell {
        
        let pair = self.data[row].parameters.object(at: subrow) as! Pair
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "SubrowCell", for: indexPath)
        cell.textLabel?.text = pair.key
        cell.detailTextLabel?.text = pair.value
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, indentationLevelForRow row: Int) -> Int {
        
        return 0
    }
    
    override func tableView(_ tableView: UITableView, indentationLevelForSubrow subrow: Int, inRow row: Int) -> Int {
        
        return 3
    }
    
    override func tableView(_ tableView: UITableView, canEditRow row: Int) -> Bool {
        
        return true
    }
    
    override func tableView(_ tableView: UITableView, canEditSubrow subrow: Int, inRow row: Int) -> Bool {
        
        return true
    }
    
    override func tableView(_ tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRow row: Int) {
        
        if editingStyle == .delete {
            
            self.data.remove(at: row)
            
            tableView.deleteRows(at: [row], with: .automatic)
            
        } else if editingStyle == .insert {
            
            
        }
    }
    
    override func tableView(_ tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forSubrow subrow: Int, inRow row: Int) {
        
        if editingStyle == .delete {
            
            self.data[row].parameters.removeObject(at: subrow)
            
            tableView.deleteSubrows([subrow], in: row, with: .automatic)
            
        } else if editingStyle == .insert {
            
            let alertController = UIAlertController(title: "Insert new subrow", placeholders: ["parameter", "value"], completion: { (texts: [String?]) -> Void in
                
                let key = texts.first! ?? "??"
                let value = texts.last! ?? "??"
                
                self.data[row].parameters.insert(Pair(key: key, value: value), at: 0)
                
                tableView.insertSubrows([0], in: row, with: .automatic)
            })
            
            self.present(alertController, animated: true, completion: nil)
        }
    }
    
    override func tableView(_ tableView: UITableView, editingStyleForRow row: Int) -> UITableViewCellEditingStyle {
        
        return .delete
    }
    
    override func tableView(_ tableView: UITableView, editingStyleForSubrow subrow: Int, inRow row: Int) -> UITableViewCellEditingStyle {
        
        return .delete
    }
    
    override func tableView(_ tableView: UITableView, didEndDisplayingCell cell: UITableViewCell, forRow row: Int) {
        
        //print(cell.textLabel?.text ?? "", row)
    }
    
    override func tableView(_ tableView: UITableView, didEndDisplayingCell cell: UITableViewCell, forSubrow subrow: Int, inRow row: Int) {
        
        //print(cell.detailTextLabel?.text ?? "", row, subrow)
    }
    
    override func tableView(_ tableView: UITableView, canMoveRow row: Int) -> Bool {
        
        return true
    }
    
    override func tableView(_ tableView: UITableView, canMoveSubrow subrow: Int, inRow row: Int) -> Bool {
        
        return true
    }
    
    override func tableView(_ tableView: UITableView, targetRowForMoveFromRow sourceRow: Int, to proposedDestinationRow: Int) -> Int {
        
        return proposedDestinationRow
    }
    
    override func tableView(_ tableView: UITableView, targetSubrowForMoveFromSubrow sourceSubrow: Int, to proposedDestinationSubrow: Int, inRow row: Int) -> Int {
        
        return proposedDestinationSubrow
    }
    
    override func tableView(_ tableView: UITableView, moveRow sourceRow: Int, to destinationRow: Int) {
        
        swap(&self.data[sourceRow], &self.data[destinationRow])
    }
    
    override func tableView(_ tableView: UITableView, moveSubrow sourceSubrow: Int, to destinationSubrow: Int, inRow row: Int) {
        
        let state = self.data[row]
        state.parameters.exchangeObject(at: sourceSubrow, withObjectAt: destinationSubrow)
    }
    
    override func tableView(_ tableView: UITableView, animationForInsertionIn row: Int) -> UITableViewRowAnimation {
        
        return .right
    }
    
    override func tableView(_ tableView: UITableView, animationForDeletionIn row: Int) -> UITableViewRowAnimation {
        
        return .left
    }
}
